### PR TITLE
feat: [FRONTEND-3500] added async validation pending status changes

### PR DESCRIPTION
**Pull request by shravyaRB**
# Description

Mentioned in https://docs.google.com/document/d/1p0zVmAuc0u2ksMRWT-WaqXg3obtFlG1cqhIOk15OQW0/edit#
 one-ui PR: https://github.com/Rippling/rippling-one-ui/pull/1605

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -18,7 +18,8 @@
     "@typescript-eslint/no-explicit-any": 0,
     "import/no-extraneous-dependencies": ["error", {"devDependencies": ["setupTests.ts", "**/*spec.tsx"]}],
     "prettier/prettier": ["error", { "singleQuote": true }],
-    "react/forbid-prop-types": 0
+    "react/forbid-prop-types": 0,
+    "react/destructuring-assignment": 0,
```
Comment (shravyaRB) [inline .eslintrc:22]: to allow using state and prop variables without destructuring

File: src/Wrapper.ts [hunk 2]
```diff
@@ -185,7 +209,7 @@ export default function<T>(
     }
 
     public componentDidUpdate(prevProps) {
-      const { value, validations, required: _required, isRequired } = this.props;
```
Comment (shravyaRB) [inline src/Wrapper.ts:188]: unused

File: src/Wrapper.ts [hunk 3]
```diff
@@ -241,6 +271,7 @@ export default function<T>(
           {
             value,
             isPristine,
+            isValidationInProgress: hasAsyncValidation,
```
Comment (shravyaRB) [inline src/Wrapper.ts:274]: conditionally set `isValidationInProgress` only when input has async validations. This will make sure not to disrupt existing sync validations flow

File: src/index.ts [hunk 4]
```diff
@@ -477,28 +495,43 @@ class Formsy extends React.Component<FormsyProps, FormsyState> {
...
+            validationError: validation.error,
+            isValidationInProgress: validation.isValidationInProgress,
+          },
+          () => {
+            const isSyncInputWithAsynValidatorsOnForm =
+              !component.state.hasAsyncValidation && Object.values(this.hasAsyncInputsValidation).some(Boolean);
+            // If input is invalid avoid waiting until Form is async validated. This is done to avoid false clicks
+            if (isSyncInputWithAsynValidatorsOnForm && !component.state.isValid) {
+              this.setFormValidState(false);
+            }
```
Comment (shravyaRB) [inline src/index.ts:520]: newly added

File: src/index.ts [hunk 5]
```diff
@@ -477,28 +495,43 @@ class Formsy extends React.Component<FormsyProps, FormsyState> {
...
-          isRequired: validation.isRequired,
-          isValid: validation.isValid,
-          validationError: validation.error,
-        },
-        this.validateForm,
-      );
-    });
+    if (component.state.hasAsyncValidation) {
+      this.setFormValidState(false);
+    }
```
Comment (shravyaRB) [inline src/index.ts:500]: newly added

File: src/index.ts [hunk 6]
```diff
@@ -477,28 +495,43 @@ class Formsy extends React.Component<FormsyProps, FormsyState> {
...
+              !component.state.hasAsyncValidation && Object.values(this.hasAsyncInputsValidation).some(Boolean);
+            // If input is invalid avoid waiting until Form is async validated. This is done to avoid false clicks
+            if (isSyncInputWithAsynValidatorsOnForm && !component.state.isValid) {
+              this.setFormValidState(false);
+            }
+            this.validateForm(); // Run validation on all in case there are dependant inputs
+          },
+        );
+      })
+      .catch(() => {});
```
Comment (shravyaRB) [inline src/index.ts:525]: added catch block now that we reject earlier requests

File: src/index.ts (hunk 4 above)
Comment (shravyaRB) [inline src/index.ts:520]: newly added

File: src/index.ts [hunk 7]
```diff
@@ -508,26 +541,36 @@ class Formsy extends React.Component<FormsyProps, FormsyState> {
   public validateForm = () => {
     // Run validation again in case affected by other inputs.
     // last component validated will run the onValidationComplete callback
-    this.runValidationOnAllInputs().then(validationResults => {
```
Comment (shravyaRB) [inline src/index.ts:511]: format changes after adding catch block

File: src/index.ts [hunk 8]
```diff
@@ -508,26 +541,36 @@ class Formsy extends React.Component<FormsyProps, FormsyState> {
...
-            },
-            index === this.inputs.length - 1 ? this.onValidationComplete : null,
-          );
+    this.latestFormValidationCounter += 1;
+    const token = this.latestFormValidationCounter;
+    this.runValidationOnAllInputs()
+      .then(validationResults => {
+        // stale promise should not do setState operations
+        if (token !== this.latestFormValidationCounter) {
+          return;
```
Comment (shravyaRB) [inline src/index.ts:550]: newly added

File: src/index.ts [hunk 9]
```diff
@@ -216,7 +223,7 @@ class Formsy extends React.Component<FormsyProps, FormsyState> {
   };
 
   // We need this callback as we are validating all inputs again and it is run when the last component has set its state
-  onValidationComplete = () => {
+  public onValidationComplete = () => {
```
Comment (shravyaRB) [inline src/index.ts:226]: lint error fix

File: src/index.ts [hunk 10]
```diff
@@ -327,21 +333,30 @@ class Formsy extends React.Component<FormsyProps, FormsyState> {
   };
 
   // Checks validation on current value or a passed value
-  public runValidation = (component: InputComponent, value = component.state.value) => {
+  public runValidation = (
+    component: InputComponent,
+    value = component.state.value,
+    setOnValidationProgress?: Function,
```
Comment (shravyaRB) [inline src/index.ts:339]: callback fn triggered before running validations in `utils.runRules`  to give info to input whether it needs to set the state for pending states for async validation

File: src/index.ts [hunk 11]
```diff
@@ -465,6 +480,13 @@ class Formsy extends React.Component<FormsyProps, FormsyState> {
     }
   };
 
+  public setInputHasAsyncValidation = (component: InputComponent, hasAsyncValidation: boolean) => {
+    this.hasAsyncInputsValidation = {
+      ...this.hasAsyncInputsValidation,
+      [component.props.name]: hasAsyncValidation,
+    };
```
Comment (shravyaRB) [inline src/index.ts:487]: this is maintained for below
Screenshot 2022-02-17 at 1 32 24 PM

File: src/index.ts [hunk 12]
```diff
@@ -477,28 +499,53 @@ class Formsy extends React.Component<FormsyProps, FormsyState> {
...
+      component.setState({
+        hasAsyncValidation: isAsyncValidation,
+        isValidationInProgress: isAsyncValidation,
+      });
+      this.setInputHasAsyncValidation(component, isAsyncValidation);
+      // disable form till validations are resolved
+      if (isAsyncValidation) {
+        this.setFormValidState(false);
+      }
+    };
```
Comment (shravyaRB) [inline src/index.ts:514]: callback triggered from https://github.com/Rippling/formsy-react/blob/41653decf8ec8c988e13dcf97f0e905d54dfa0fc/src/utils.ts#L118toL124

File: src/index.ts [hunk 13]
```diff
@@ -508,26 +555,36 @@ class Formsy extends React.Component<FormsyProps, FormsyState> {
...
+                externalError:
+                  !validation.isValid && component.state.externalError ? component.state.externalError : null,
+                isValidationInProgress: false,
+              },
+              index === this.inputs.length - 1 ? this.onValidationComplete : null,
+            );
+          }
+        });
+      })
+      .catch(() => {});
```
Comment (shravyaRB) [inline src/index.ts:587]: format changes after adding catch block

File: src/index.ts [hunk 14]
```diff
@@ -327,21 +333,30 @@ class Formsy extends React.Component<FormsyProps, FormsyState> {
...
-      utils.runRules(value, currentValues, component.validations, validationRules),
+      utils.runRules(value, currentValues, component.validations, validationRules, setOnValidationProgress),
       utils.runRules(value, currentValues, component.requiredValidations, validationRules),
     ]).then(([validationResults, requiredResults]) => {
       const isRequired = Object.keys(component.requiredValidations).length ? !!requiredResults.success.length : false;
       const isValid = !validationResults.failed.length && !(validationErrors && validationErrors[component.props.name]);
+      // reject stale promises
+      if (!utils.isSame(value, component.state.value)) {
+        return Promise.reject();
+      }
```
Comment (mrdivyansh) [inline src/index.ts:355]: @shravyaRB what is reasoning behind this?

File: src/index.ts [hunk 15]
```diff
@@ -477,28 +499,53 @@ class Formsy extends React.Component<FormsyProps, FormsyState> {
...
   /**
    * returns validations array after async validation on each input is resolved
    */
-  runValidationOnAllInputs = () => {
+  public runValidationOnAllInputs = () => {
     const validationPromises = [];
-    this.inputs.forEach((component, index) => {
-      validationPromises.push(this.runValidation(component));
+    this.inputs.forEach(component => {
+      validationPromises.push(this.runValidation(component, component.state.value));
```
Comment (mrdivyansh) [inline src/index.ts:548]: Why this change?

File: src/index.ts (hunk 14 above)
Comment (shravyaRB) [reply src/index.ts:355]: adding console.log to show demo

https://www.loom.com/share/23e4164e7d934bf489074538e489a8cb

File: src/index.ts [hunk 16]
```diff
@@ -158,6 +164,7 @@ class Formsy extends React.Component<FormsyProps, FormsyState> {
       isFormDisabled: this.isFormDisabled(),
       isValidValue: this.isValidValue,
       validate: this.validate,
+      setInputHasAsyncValidation: this.setInputHasAsyncValidation,
```
Comment (mrdivyansh) [inline src/index.ts:167]: Remove

File: src/index.ts [hunk 17]
```diff
@@ -75,6 +75,10 @@ class Formsy extends React.Component<FormsyProps, FormsyState> {
 
   public prevInputNames: any[] | null = null;
 
+  public hasAsyncInputsValidation: Record<string, boolean>;
```
Comment (mrdivyansh) [inline src/index.ts:78]: Suffix with map

File: src/index.ts [hunk 18]
```diff
@@ -477,28 +499,53 @@ class Formsy extends React.Component<FormsyProps, FormsyState> {
...
+            isValidationInProgress: validation.isValidationInProgress,
+          },
+          () => {
+            const isSyncInputWithAsynValidatorsOnForm =
+              !component.state.hasAsyncValidation && Object.values(this.hasAsyncInputsValidation).some(Boolean);
+            // If input is invalid avoid waiting until Form is async validated. This is done to avoid false clicks on save btn
+            if (isSyncInputWithAsynValidatorsOnForm && !component.state.isValid) {
+              this.setFormValidState(false);
+            }
+            this.validateForm(); // Run validation on all in case there are dependant inputs
```
Comment (mrdivyansh) [inline src/index.ts:535]: Seems the field validation will be stuck in a loop.

Comment (shravyaRB) [reply src/index.ts:535]: `validateForm` is not calling this function. It directly calls `runValidation` function
